### PR TITLE
Homework 18 was done. Generators, iterators and decorators where impl…

### PR DESCRIPTION
**Pull request by 213140**
…emented in separate files and checked

### CODE REVIEW

File: lesson_18/homework_18_iterators.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.current_position <= 0 & self.current_position < self.initial_list.__len__():
+            self.current_position -= 1
+            return self.initial_list[self.current_position]
+        else:
+            return StopIteration
```
Comment (edanilovets) [inline lesson_18/homework_18_iterators.py]: when there are no elements left the re should be StopIteration exception
When I ran your code I got
```
Traceback (most recent call last):
  File "/Users/danilovets/Personal/Hillel/temp/AQA_Python_13082024/lesson_18/homework_18_iterators.py", line 48, in <module>
    print(next(iter1))
          ^^^^^^^^^^^
  File "/Users/danilovets/Personal/Hillel/temp/AQA_Python_13082024/lesson_18/homework_18_iterators.py", line 17, in __next__
    return self.initial_list[self.current_position]
           ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
IndexError: list index out of range
```

File: lesson_18/homework_18_decorators.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@logging_decorator
+def some_function(city_name: str, city_residents: int):
+    print(f"City name is {city_name} and has {city_residents} residents")
+# Check if decorator works
+some_function('Warsaw', 800000)
+
+# 2 way of using
+def some_function(city_name: str, city_residents: int):
+     print(f"City name is {city_name} and has {city_residents} residents")
+# Crete copy of original function
```
Comment (edanilovets) [inline lesson_18/homework_18_decorators.py]: please follow pep8 standard for code stylings, too many style erros
Screenshot 2024-10-18 at 12 53 19

File: lesson_18/homework_18_decorators.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+copy_of_function = some_function
+# Connect copy with decorator
+dec2_with_func = logging_decorator(copy_of_function)
+# Call both: copy with deco. and original function
+dec2_with_func('Warsaw', 800000)
+copy_of_function('Warsaw', 800000)
+
+
+
```
Comment (edanilovets) [inline lesson_18/homework_18_decorators.py]: there was 1 more task
Створіть декоратор, який перехоплює та обробляє винятки, які виникають в ході виконання функції.

File: lesson_18/homework_18_iterators.py (hunk 1 above)
Comment (213140) [reply lesson_18/homework_18_iterators.py]: expresion was updated. Need to be checked

File: lesson_18/homework_18_decorators.py (hunk 3 above)
Comment (213140) [reply lesson_18/homework_18_decorators.py]: Missing decorato was added. Need to be verified by reviewer.

**Review by edanilovets — APPROVED** [review]

**Pull request merged**